src/docfx/build/moniker/MonikerProvider.cs

```diff
@@ -28,28 +34,47 @@ public MonikerProvider(Docset docset)
 
             foreach (var (key, monikerRange) in docset.Config.MonikerRange)
             {
-                _rules.Add((GlobUtility.CreateGlobMatcher(key), (monikerRange, _rangeParser.ParseWithInfo(monikerRange))));
+                _rules.Add((GlobUtility.CreateGlobMatcher(key), (monikerRange, _rangeParser.Parse(monikerRange))));
             }
             _rules.Reverse();
         }
 
-        public (List<Error> errors, List<string> monikers) GetFileLevelMonikers(Document file, MetadataProvider metadataProvider)
+        public (Error error, List<string> monikers) GetFileLevelMonikers(Document file)
         {
-            var errors = new List<Error>();
-            var (metaErrors, metadata) = metadataProvider.GetInputMetadata<InputMetadata>(file, null);
-            errors.AddRange(metaErrors);
+            return _monikerCache.GetOrAdd(file, GetFileLevelMonikersCore);
+        }
+
+        public (Error error, List<string> monikers) GetZoneLevelMonikers(Document file, SourceInfo<string> rangeString)
+        {
+            var (_, fileLevelMonikers) = GetFileLevelMonikers(file);
+
+            // Moniker range not defined in docfx.yml/docfx.json,
+            // User should not define it in moniker zone
+            if (fileLevelMonikers.Count == 0)
+            {
+                return (Errors.MonikerConfigMissing(), new List<string>());
+            }
 
-            var (monikerError, monikers) = GetFileLevelMonikers(file, metadata.MonikerRange);
-            errors.AddIfNotNull(monikerError);
+            var zoneLevelMonikers = _rangeParser.Parse(rangeString);
+            var monikers = fileLevelMonikers.Intersect(zoneLevelMonikers, StringComparer.OrdinalIgnoreCase).ToList();
 
-            return (errors, monikers);
+            if (monikers.Count == 0)
+            {
+                var error = Errors.EmptyMonikers($"No intersection between zone and file level monikers. The result of zone level range string '{rangeString}' is '{string.Join(',', zoneLevelMonikers)}', while file level monikers is '{string.Join(',', fileLevelMonikers)}'.");
```

Comment:
Do we have criteria when we should pass an error message and when it should be prepared within the `Error` itself?
Because with this way, it would be harder for us to gather the error messages for PM to review

Reply:
this is a special case, because we want to use the same code `empty-moniker` for the config/file and file/zone empty moniker intersection, but with different message, so we pass an error message here.

Reply:
The message can still be prepared

Reply:
@928PJY will you do this refactor later?

Reply:
OK, leave it to me.